File: backend/app/schemas/supplier.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict


# --- Supplier ---

class SupplierBase(BaseModel):
    name: str
    description: Optional[str] = None
    website: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    address: Optional[str] = None
    notes: Optional[str] = None
    is_active: bool = True


class SupplierCreate(SupplierBase):
    pass


class SupplierUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    website: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    address: Optional[str] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class SupplierResponse(SupplierBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class SupplierListResponse(BaseModel):
    items: list[SupplierResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# --- Manufacturer ---

class ManufacturerBase(BaseModel):
    name: str
    description: Optional[str] = None
    website: Optional[str] = None
    notes: Optional[str] = None
    is_active: bool = True


class ManufacturerCreate(ManufacturerBase):
    pass


class ManufacturerUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    website: Optional[str] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class ManufacturerResponse(ManufacturerBase):
    id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class ManufacturerListResponse(BaseModel):
    items: list[ManufacturerResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# --- SupplierPart ---

class SupplierPartBase(BaseModel):
    component_id: int
    supplier_id: int
    supplier_part_number: Optional[str] = None
    unit_price: Optional[float] = None
    currency: str = "USD"
    pack_quantity: int = 1
    lead_time_days: Optional[int] = None
    url: Optional[str] = None
    notes: Optional[str] = None
    is_active: bool = True


class SupplierPartCreate(BaseModel):
    component_id: int
    supplier_id: Optional[int] = None
    supplier_part_number: Optional[str] = None
    unit_price: Optional[float] = None
    currency: str = "USD"
    pack_quantity: int = 1
    lead_time_days: Optional[int] = None
    url: Optional[str] = None
    notes: Optional[str] = None
    is_active: bool = True


class SupplierPartUpdate(BaseModel):
    supplier_part_number: Optional[str] = None
    unit_price: Optional[float] = None
    currency: Optional[str] = None
    pack_quantity: Optional[int] = None
    lead_time_days: Optional[int] = None
    url: Optional[str] = None
    notes: Optional[str] = None
    is_active: Optional[bool] = None


class SupplierPartResponse(SupplierPartBase):
    id: int
    supplier_name: Optional[str] = None
    component_name: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


# --- ManufacturerPart ---

class ManufacturerPartBase(BaseModel):
    component_id: int
    manufacturer_id: int
    manufacturer_part_number: str
    description: Optional[str] = None
    url: Optional[str] = None


class ManufacturerPartCreate(BaseModel):
    component_id: int
    manufacturer_id: Optional[int] = None
    manufacturer_part_number: str
    description: Optional[str] = None
    url: Optional[str] = None


class ManufacturerPartUpdate(BaseModel):
    manufacturer_part_number: Optional[str] = None
    description: Optional[str] = None
    url: Optional[str] = None


class ManufacturerPartResponse(ManufacturerPartBase):
    id: int
    manufacturer_name: Optional[str] = None
    component_name: Optional[str] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/supplier.ts
export interface Supplier {
  id: number;
  name: string;
  description: string | null;
  website: string | null;
  phone: string | null;
  email: string | null;
  address: string | null;
  notes: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string | null;
}

export interface SupplierCreate {
  name: string;
  description?: string;
  website?: string;
  phone?: string;
  email?: string;
  address?: string;
  notes?: string;
}

export interface SupplierUpdate extends Partial<SupplierCreate> {
  is_active?: boolean;
}

export interface Manufacturer {
  id: number;
  name: string;
  description: string | null;
  website: string | null;
  notes: string | null;
  is_active: boolean;
  created_at: string;
  updated_at: string | null;
}

export interface ManufacturerCreate {
  name: string;
  description?: string;
  website?: string;
  notes?: string;
}

export interface ManufacturerUpdate extends Partial<ManufacturerCreate> {
  is_active?: boolean;
}

export interface SupplierPart {
  id: number;
  component_id: number;
  supplier_id: number;
  supplier_part_number: string | null;
  unit_price: number | null;
  currency: string;
  pack_quantity: number;
  lead_time_days: number | null;
  url: string | null;
  notes: string | null;
  is_active: boolean;
  supplier_name: string;
  component_name: string;
  created_at: string;
}

export interface ManufacturerPart {
  id: number;
  component_id: number;
  manufacturer_id: number;
  manufacturer_part_number: string;
  description: string | null;
  url: string | null;
  manufacturer_name: string;
  component_name: string;
  created_at: string;
}
